src/modules/flow/json/json.c

```diff
@@ -0,0 +1,221 @@
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "sol-flow/json.h"
+#include "sol-flow-internal.h"
+
+#include <sol-json.h>
+#include <sol-util.h>
+#include <sol-types.h>
+#include <errno.h>
+
+static const struct sol_blob_type sub_json_blob_type = {
+    .api_version = SOL_BLOB_TYPE_API_VERSION,
+};
+
+struct sol_json_object_key {
+    struct sol_blob *json_object;
+    char *key;
+};
+
+static int
+json_object_get_key_open(struct sol_flow_node *node, void *data, const struct sol_flow_node_options *options)
+{
+    struct sol_json_object_key *mdata = data;
+    const struct sol_flow_node_type_json_object_get_key_options *opts;
+
+    SOL_FLOW_NODE_OPTIONS_SUB_API_CHECK(options,
+        SOL_FLOW_NODE_TYPE_JSON_OBJECT_GET_KEY_OPTIONS_API_VERSION,
+        -EINVAL);
+    opts = (const struct sol_flow_node_type_json_object_get_key_options *)
+        options;
+
+    mdata->key = strdup(opts->key);
+    SOL_NULL_CHECK(mdata->key, -ENOMEM);
+
+    return 0;
+}
+
+static void
+json_object_get_key_close(struct sol_flow_node *node, void *data)
+{
+    struct sol_json_object_key *mdata = data;
+
+    if (mdata->json_object)
+        sol_blob_unref(mdata->json_object);
+    free(mdata->key);
+}
+
+static struct sol_blob *
+create_sub_json_object(struct sol_blob *parent, struct sol_json_token *token)
+{
+    struct sol_json_scanner scanner;
+    const char *mem;
+
+    mem = token->start;
+    sol_json_scanner_init(&scanner, parent->mem, parent->size);
+    if (!sol_json_scanner_skip_over(&scanner, token))
+        return NULL;
+
+    if (sol_json_token_get_type(token) != SOL_JSON_TYPE_OBJECT_END) {
+        errno = EINVAL;
+        return NULL;
+    }
+
+    return sol_blob_new(&sub_json_blob_type, parent, mem, token->end - mem);
+}
+
+static int
+json_object_key_process(struct sol_flow_node *node, struct sol_json_object_key *mdata)
+{
+    struct sol_json_scanner scanner;
+    struct sol_json_token token, key, value;
+    enum sol_json_loop_reason reason;
+    enum sol_json_type type;
+    unsigned int len;
+    struct sol_irange value_int = SOL_IRANGE_INIT();
+    struct sol_drange value_float = SOL_DRANGE_INIT();
+    struct sol_str_slice slice;
+    struct sol_blob *new_blob;
+    int r;
+
+    if (!mdata->key[0] || !mdata->json_object)
+        return 0;
+
+    len = strlen(mdata->key);
+    sol_json_scanner_init(&scanner, mdata->json_object->mem,
+        mdata->json_object->size);
+    SOL_JSON_SCANNER_OBJECT_LOOP (&scanner, &token, &key, &value, reason) {
+        if (sol_json_token_str_eq(&key, mdata->key, len)) {
```

Comment:
you're not handling JSON Path here and it's okay, but you should prepare your code to.

To do so, make this whole loop a function and give it the path. Then when you want to add the JSON Path variant, what you can do is extract the pieces from the path (ie: skip `/`, get the key up to the next `/` or `\0`, then recurse with the remaining key slice).

You can also do an initial check if `len == 0`, there is no key and you can return as nothing will be found.